{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Standard variables loaded, you are good to go!\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from dotenv import load_dotenv, find_dotenv\n",
    "from os.path import join, dirname, basename, exists, isdir\n",
    "\n",
    "### Load environmental variables from the project root directory ###\n",
    "# find .env automagically by walking up directories until it's found\n",
    "dotenv_path = find_dotenv()\n",
    "\n",
    "# load up the entries as environment variables\n",
    "load_dotenv(dotenv_path)\n",
    "\n",
    "# now you can get the variables using their names\n",
    "\n",
    "# Check whether a network drive has been specified\n",
    "DATABASE = os.environ.get(\"NETWORK_URL\")\n",
    "if DATABASE == 'None':\n",
    "    pass\n",
    "else:\n",
    "    pass\n",
    "    #mount network drive here\n",
    "\n",
    "# set up directory paths\n",
    "CURRENT_DIR = os.getcwd()\n",
    "PROJ = dirname(dotenv_path) # project root directory\n",
    "\n",
    "DATA = join(PROJ, 'data') #data directory\n",
    "RAW_EXTERNAL = join(DATA, 'raw_external') # external data raw directory\n",
    "RAW_INTERNAL = join(DATA, 'raw_internal') # internal data raw directory\n",
    "INTERMEDIATE = join(DATA, 'intermediate') # intermediate data directory\n",
    "FINAL = join(DATA, 'final') # final data directory\n",
    "\n",
    "RESULTS = join(PROJ, 'results') # output directory\n",
    "FIGURES = join(RESULTS, 'figures') # figure output directory\n",
    "PICTURES = join(RESULTS, 'pictures') # picture output directory\n",
    "\n",
    "\n",
    "# make folders specific for certain data\n",
    "folder_name = 'BRENDA_2019.1_ec_data'\n",
    "if folder_name != '':\n",
    "    #make folders if they don't exist\n",
    "    if not exists(join(RAW_EXTERNAL, folder_name)):\n",
    "        os.makedirs(join(RAW_EXTERNAL, folder_name))\n",
    "\n",
    "    if not exists(join(INTERMEDIATE, folder_name)):\n",
    "        os.makedirs(join(INTERMEDIATE, folder_name))\n",
    "\n",
    "    if not exists(join(FINAL, folder_name)):\n",
    "        os.makedirs(join(FINAL, folder_name))\n",
    "\n",
    "print('Standard variables loaded, you are good to go!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import pandas as pd\n",
    "import time\n",
    "import numpy as np\n",
    "from collections import Counter, defaultdict\n",
    "\n",
    "\n",
    "seq = \"MAVSAGSARTSPSSDKVQKDKAELISGPRQDSRIGKLLGFEWTDLSSWRRLVTLLNRPTDPASLAVFRFLFGFLMVLDIPQERGLSSLDRKYLDGLDVCRFPLLDALRPLPLDWMYLVYTIMFLGALGMMLGLCYRISCVLFLLPYWYVFLLDKTSWNNHSYLYGLLAFQLTFMDANHYWSVDGLLNAHRRNAHVPLWNYAVLRGQIFIVYFIAGVKKLDADWVEGYSMEYLSRHWLFSPFKLLLSEELTSLLVVHWGGLLLDLSAGFLLFFDVSRSIGLFFVSYFHCMNSQLFSIGMFSYVMLASSPLFCSPEWPRKLVSYCPRRLQQLLPLKAAPQPSVSCVYKRSRGKSGQKPGLRHQLGAAFTLLYLLEQLFLPYSHFLTQGYNNWTNGLYGYSWDMMVHSRSHQHVKITYRDGRTGELGYLNPGVFTQSRRWKDHADMLKQYATCLSRLLPKYNVTEPQIYFDIWVSINDRFQQRIFDPRVDIVQAAWSPFQRTSWVQPLLMDLSPWRAKLQEIKSSLDNHTEVVFIADFPGLHLENFVSEDLGNTSIQLLQGEVTVELVAEQKNQTLREGEKMQLPAGEYHKVYTTSPSPSCYMYVYVNTTELALEQDLAYLQELKEKVENGSETGPLPPELQPLLEGEVKGGPEPTPLVQTFLRRQQRLQEIERRRNTPFHERFFRFLLRKLYVFRRSFLMTCISLRNLILGRPSLEQLAQEVTYANLRPFEAVGELNPSNTDSSHSNPPESNPDPVHSEF\"\n",
    "\n",
    "# get the vector data\n",
    "filename = join(RAW_EXTERNAL, 'protvec', 'protVec_100d_3grams.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_data(filepath):\n",
    "    '''\n",
    "    Load up the vector data\n",
    "    '''\n",
    "    return pd.read_csv(filepath, header=None, sep='\\t')\n",
    "\n",
    "\n",
    "\n",
    "def round_number(x, factor=100):\n",
    "    '''\n",
    "    Multiply number by a factor and then convert to an int\n",
    "    '''\n",
    "    return int(round(x * factor))\n",
    "\n",
    "\n",
    "\n",
    "def score_seq_original(seq, enbedding, to_int=True):\n",
    "    '''\n",
    "    Use the embedding to calculate a feature vector for a protein sequence\n",
    "    '''\n",
    "    seq = seq.upper()\n",
    "    \n",
    "    # make all the three-mers\n",
    "    three_mers = [seq[i:i+3] for i in range(0, len(seq)-2)]\n",
    "\n",
    "    # convert the sequence 3-mer list to a data frame\n",
    "    three_mers_df = pd.DataFrame()\n",
    "    three_mers_df[0] = three_mers\n",
    "\n",
    "    # merge the two data frames, the merge happens on the sequence trimers\n",
    "    merged_df = pd.merge(three_mers_df, enbedding, on=[0])\n",
    "\n",
    "    # sum the columns and remove the sequence information (first column)\n",
    "    vector_sums = merged_df.sum(axis = 0, skipna = False).drop(0)\n",
    "    \n",
    "    # to save on memory we may want to convert the folat to an int\n",
    "    if to_int is True:\n",
    "        vector_sums = vector_sums.apply(round_number)\n",
    "    \n",
    "    return vector_sums.tolist()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "    \n",
    "# load up the enbedding data\n",
    "enbedding_df = load_data(filename)   \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "n = 100\n",
    "\n",
    "t1 = time.time()\n",
    "\n",
    "for i in range(n):\n",
    "    vector = score_seq_original(seq, enbedding_df, to_int=False)\n",
    "t2 = time.time()\n",
    "\n",
    "print('time per seq: ', (t2-t1)/n)\n",
    "\n",
    "print(vector)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def round_number(x, factor=100):\n",
    "    '''\n",
    "    Multiply number by a factor and then convert to an int\n",
    "    '''\n",
    "    return int(round(x * factor))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_data(filename):\n",
    "    '''\n",
    "    Load up the vector data\n",
    "    '''\n",
    "    return pd.read_csv(filename, header=None, sep='\\t',index_col=0)\n",
    "\n",
    "\n",
    "def score_seq_gang(seq, enbedding, to_int=True):\n",
    "    '''\n",
    "    Use the embedding to calculate a feature vector for a protein sequence\n",
    "    '''\n",
    "    seq = seq.upper()\n",
    "    \n",
    "    # make all the three-mers\n",
    "    three_mers = dict()\n",
    "    for i in range(0, len(seq)-2): \n",
    "        kmer = seq[i:i+3]\n",
    "        three_mers[kmer] = three_mers.get(kmer, 0) + 1\n",
    "        \n",
    "    c = [three_mers.get(kmer, 0) for kmer in enbedding.index]\n",
    "    \n",
    "    vector_sums = np.dot(np.array(c), enbedding)\n",
    "    \n",
    "    if to_int is True:\n",
    "        vector_sums = vector_sums.apply(round_number)\n",
    "    \n",
    "    return vector_sums.tolist()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# load up the enbedding data\n",
    "enbedding_df = load_data(filename)   \n",
    "\n",
    "\n",
    "# calculate an enbedded representation of the protein\n",
    "t1 = time.time()\n",
    "\n",
    "n = 1000\n",
    "\n",
    "for i in range(n):\n",
    "       \n",
    "    vector = score_seq_gang(seq, enbedding_df, to_int=False)\n",
    "    \n",
    "t2 = time.time()\n",
    "\n",
    "print('time per seq: %s \\n\\n' % ((t2-t1)/n))\n",
    "\n",
    "print(vector)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_data_alternate(filename):\n",
    "    '''\n",
    "    Load up the vector data\n",
    "    '''\n",
    "    data = {}\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            elements = line.split('\\t')\n",
    "            data[elements[0]] = np.array([float(i) for i in elements[1:]])\n",
    "    return data\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def score_seq_4(seq, enbedding, to_int=True):\n",
    "    '''\n",
    "    Use the embedding to calculate a feature vector for a protein sequence.\n",
    "    \n",
    "    This is the fastest one. Gives a performance of about 4.2 hours for 10 mio sequences.\n",
    "    '''\n",
    "    seq = seq.upper()  \n",
    "    \n",
    "    # make all the three-mers\n",
    "    three_mers = Counter([seq[i:i+3] for i in range(0, len(seq)-2)])\n",
    "    \n",
    "    # multiply the occurances with the feature vector\n",
    "    c = [enbedding[k] * v for k, v in three_mers.items()]\n",
    "    \n",
    "    # sum the columns of the generated vectors\n",
    "    vector_sums = np.array(c).sum(axis=0)\n",
    "    \n",
    "    # convert to int if desired\n",
    "    if to_int is True:\n",
    "        vector_sums = vector_sums.apply(round_number)\n",
    "    \n",
    "    return vector_sums.tolist()\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# calculate an enbedded representation of the protein\n",
    "\n",
    "dict_enbedding = load_data_alternate(filename)\n",
    "\n",
    "\n",
    "t1 = time.time()\n",
    "\n",
    "n = 1000\n",
    "\n",
    "for i in range(n):\n",
    "        \n",
    "       \n",
    "    vector = score_seq_4(seq, dict_enbedding, to_int=False)\n",
    "    \n",
    "t2 = time.time()\n",
    "\n",
    "print('time per seq: %s \\n\\n' % ((t2-t1)/n))\n",
    "\n",
    "print(vector)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_data_alternate_2(filename):\n",
    "    '''\n",
    "    Load up the vector data\n",
    "    '''\n",
    "    data = {}\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            elements = line.split('\\t')\n",
    "            \n",
    "            if elements[0] == '<unk>':\n",
    "                continue\n",
    "            \n",
    "            one, two, three = elements[0]\n",
    "            \n",
    "            if data.get(one) is None:\n",
    "                data[one] = {}\n",
    "                \n",
    "            if data[one].get(two) is None:\n",
    "                data[one][two] = {}\n",
    "                \n",
    "            if data[one][two].get(three) is None:\n",
    "                data[one][two][three] = {}\n",
    "                \n",
    "            for num in range(1, 10):\n",
    "                data[one][two][three][num] = np.array([float(i)*num for i in elements[1:]])\n",
    "                \n",
    "    return data\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def score_seq_5(seq, enbedding, to_int=True):\n",
    "    '''\n",
    "    Use the embedding to calculate a feature vector for a protein sequence.\n",
    "    \n",
    "    This is the fastest one. Gives a performance of about 4.2 hours for 10 mio sequences.\n",
    "    '''\n",
    "    seq = seq.upper()  \n",
    "    \n",
    "    # make all the three-mers\n",
    "    three_mers = Counter([seq[i:i+3] for i in range(0, len(seq)-2)])\n",
    "    \n",
    "    # multiply the occurances with the feature vector\n",
    "    c = [enbedding[k[0]][k[1]][k[2]][v] for k, v in three_mers.items()]\n",
    "    \n",
    "    # sum the columns of the generated vectors\n",
    "    vector_sums = np.array(c).sum(axis=0)\n",
    "    \n",
    "    # convert to int if desired\n",
    "    if to_int is True:\n",
    "        vector_sums = vector_sums.apply(round_number)\n",
    "    \n",
    "    return vector_sums.tolist()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# calculate an enbedded representation of the protein\n",
    "\n",
    "dict_enbedding = load_data_alternate_2(filename)\n",
    "\n",
    "\n",
    "t1 = time.time()\n",
    "\n",
    "n = 1000\n",
    "\n",
    "for i in range(n):\n",
    "        \n",
    "       \n",
    "    vector = score_seq_5(seq, dict_enbedding, to_int=False)\n",
    "    \n",
    "t2 = time.time()\n",
    "\n",
    "print('time per seq: %s \\n\\n' % ((t2-t1)/n))\n",
    "\n",
    "print(vector)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "def load_data_alternate_3(filename):\n",
    "    '''\n",
    "    Load up the vector data\n",
    "    '''\n",
    "    data = {}\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            elements = line.split('\\t')\n",
    "            \n",
    "            if elements[0] == '<unk>':\n",
    "                continue\n",
    "            \n",
    "            codon = elements[0]\n",
    "                \n",
    "            data[codon] = {}\n",
    "                \n",
    "            for num in range(1, 10):\n",
    "                data[codon][num] = np.array([float(i)*num for i in elements[1:]])\n",
    "                \n",
    "    return data\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def score_seq_6(seq, enbedding, to_int=True):\n",
    "    '''\n",
    "    Use the embedding to calculate a feature vector for a protein sequence.\n",
    "    \n",
    "    This is the fastest one. Gives a performance of about 4.2 hours for 10 mio sequences.\n",
    "    '''\n",
    "    seq = seq.upper()  \n",
    "    \n",
    "    # make all the three-mers\n",
    "    three_mers = Counter([seq[i:i+3] for i in range(0, len(seq)-2)])\n",
    "    \n",
    "    # multiply the occurances with the feature vector\n",
    "    c = [enbedding[k][v] for k, v in three_mers.items()]\n",
    "    \n",
    "    # sum the columns of the generated vectors\n",
    "    vector_sums = np.array(c).sum(axis=0)\n",
    "    \n",
    "    # convert to int if desired\n",
    "    if to_int is True:\n",
    "        vector_sums = vector_sums.apply(round_number)\n",
    "    \n",
    "    return vector_sums.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# calculate an enbedded representation of the protein\n",
    "\n",
    "dict_enbedding = load_data_alternate_3(filename)\n",
    "      \n",
    "# %timeit score_seq_6(seq, dict_enbedding, to_int=False)\n",
    "\n",
    "# %prun score_seq_6(seq, dict_enbedding, to_int=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%timeit score_seq_6(seq, dict_enbedding, to_int=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# calculate an enbedded representation of the protein\n",
    "\n",
    "dict_enbedding = load_data_alternate_3(filename)\n",
    "\n",
    "\n",
    "t1 = time.time()\n",
    "\n",
    "n = 10000\n",
    "\n",
    "for i in range(n):\n",
    "        \n",
    "       \n",
    "    vector = score_seq_6(seq, dict_enbedding, to_int=False)\n",
    "    \n",
    "t2 = time.time()\n",
    "\n",
    "print('time per seq: %s \\n\\n' % ((t2-t1)/n))\n",
    "\n",
    "print(vector)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The cythonmagic extension is already loaded. To reload it, use:\n",
      "  %reload_ext cythonmagic\n"
     ]
    }
   ],
   "source": [
    "\n",
    "%load_ext cythonmagic\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "UsageError: Cell magic `%%cython` not found.\n"
     ]
    }
   ],
   "source": [
    "%%cython\n",
    "cimport cython\n",
    "from collections import Counter\n",
    "import numpy as np\n",
    "cimport numpy as np\n",
    "@cython.infer_types(True)\n",
    "cdef load_data_alternate_4(str filename):\n",
    "    '''\n",
    "    Load up the vector data\n",
    "    '''\n",
    "    cdef unsigned int num\n",
    "    cdef dict data\n",
    "    cdef str codon\n",
    "    cdef double i\n",
    "    data = {}\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            elements = line.split('\\t')\n",
    "            if elements[0] == '<unk>':\n",
    "                continue\n",
    "            codon = elements[0]\n",
    "            data[codon] = {}\n",
    "            for num in range(1, 10):\n",
    "                data[codon][num] = np.array([float(i)*num for i in elements[1:]])\n",
    "    return data\n",
    "@cython.infer_types(True)\n",
    "cdef score_seq_7(str seq, dict enbedding):\n",
    "    '''\n",
    "    Use the embedding to calculate a feature vector for a protein sequence.\n",
    "    This is the fastest one. Gives a performance of about 4.2 hours for 10 mio sequences.\n",
    "    '''\n",
    "    cdef unsigned int i\n",
    "    cdef unsigned int three = 3\n",
    "    cdef unsigned int v\n",
    "    cdef str k\n",
    "    cdef list c\n",
    "    cdef np.ndarray vector_sums\n",
    "    cdef list vector_list\n",
    "    seq = seq.upper()  \n",
    "    # make all the three-mers\n",
    "    three_mers = Counter([seq[i:i+three] for i in range(0, len(seq)-2)])\n",
    "    # multiply the occurances with the feature vector\n",
    "    c = [enbedding[k][v] for k, v in three_mers.items()]\n",
    "    # sum the columns of the generated vectors\n",
    "    vector_sums = np.array(c).sum(axis=0)\n",
    "    return vector_sums.tolist()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# calculate an enbedded representation of the protein\n",
    "\n",
    "dict_enbedding = load_data_alternate_4(filename)\n",
    "\n",
    "\n",
    "\n",
    "t1 = time.time()\n",
    "\n",
    "n = 10000\n",
    "\n",
    "for i in range(n):\n",
    "        \n",
    "       \n",
    "    vector = score_seq_7(seq, dict_enbedding)\n",
    "    \n",
    "t2 = time.time()\n",
    "\n",
    "print('time per seq: %s \\n\\n' % ((t2-t1)/n))\n",
    "\n",
    "print(vector)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_enbedding = load_data_alternate_4(filename)\n",
    "\n",
    "%timeit score_seq_7(seq, dict_enbedding)\n",
    "\n",
    "%prun score_seq_7(seq, dict_enbedding)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_enbedding\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "0.013638761043548584 / 0.0005056767463684082"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
